velox/common/compression/Compression.h

```diff
@@ -46,6 +49,218 @@ CompressionKind stringToCompressionKind(const std::string& kind);
+  virtual folly::Expected<FlushResult, Status> flush(
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+
+  /// End compressing, doing whatever is necessary to end the stream.
+  /// If EndResult.outputTooSmall is true on return, end() should be called
+  /// again with a larger buffer. Otherwise, the StreamingCompressor should not
+  /// be used anymore. end() will flush the compressed output.
+  virtual folly::Expected<EndResult, Status> end(
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+};
+
+class StreamingDecompressor {
+ public:
+  virtual ~StreamingDecompressor() = default;
+
+  struct DecompressResult {
+    uint64_t bytesRead;
+    uint64_t bytesWritten;
+    bool outputTooSmall;
+  };
+
+  /// Decompress some input.
+  /// If outputTooSmall is true on return, a larger output buffer needs
+  /// to be supplied.
+  virtual folly::Expected<DecompressResult, Status> decompress(
+      const uint8_t* input,
+      uint64_t inputLength,
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+
+  // Return whether the compressed stream is finished.
+  virtual bool isFinished() = 0;
+
+  // Reinitialize decompressor, making it ready for a new compressed stream.
+  virtual Status reset() = 0;
+};
+
+struct CodecOptions {
+  int32_t compressionLevel;
+
+  CodecOptions(int32_t compressionLevel = kUseDefaultCompressionLevel)
+      : compressionLevel(compressionLevel) {}
+
+  virtual ~CodecOptions() = default;
+};
+
+class Codec {
+ public:
+  virtual ~Codec() = default;
+
+  // Create a kind for the given compression algorithm with CodecOptions.
+  static folly::Expected<std::unique_ptr<Codec>, Status> create(
+      CompressionKind kind,
+      const CodecOptions& codecOptions = CodecOptions{});
+
+  // Create a kind for the given compression algorithm.
+  static folly::Expected<std::unique_ptr<Codec>, Status> create(
+      CompressionKind kind,
+      int32_t compressionLevel);
+
+  // Return true if support for indicated kind has been enabled.
+  static bool isAvailable(CompressionKind kind);
+
+  /// Return true if indicated kind supports extracting uncompressed length
+  /// from compressed data.
+  static bool supportsGetUncompressedLength(CompressionKind kind);
+
+  /// Return true if indicated kind supports one-shot compression with fixed
+  /// compressed length.
+  static bool supportsCompressFixedLength(CompressionKind kind);
+
+  // Return true if indicated kind supports creating streaming de/compressor.
+  static bool supportsStreamingCompression(CompressionKind kind);
+
+  /// Return the smallest supported compression level for the kind.
+  /// Note: This function creates a temporary Codec instance.
+  static folly::Expected<int32_t, Status> minimumCompressionLevel(
+      CompressionKind kind);
+
+  /// Return the largest supported compression level for the kind
+  /// Note: This function creates a temporary Codec instance.
+  static folly::Expected<int32_t, Status> maximumCompressionLevel(
```

Comment:
It is odd that these static functions have to create an instance to get the result.
The Arrow PR that added this felt there would be a usage, but I don't see this being used outside of tests.
Velox will not be exposing the compression API as a public API.
Let's move these static functions to a test utility or remove some of them that are not being used.